### DISCUSSION TITLE
Problems using CUDATookit from SDK  with CMake

### DISCUSSION THREAD

**Discussion by hannes09** [OE4T/meta-tegra#1224] [Q&A]
I have tried to use CUDAToolkit and added it to to the SDK.

`TOOLCHAIN_HOST_TASK += "nativesdk-cmake \
    nativesdk-packagegroup-cuda-sdk-host \
   nativesdk-cuda-environment"`

and 

`TOOLCHAIN_TARGET_TASK += " cuda-compiler-dev \
    cuda-cudart-dev \
    cuda-toolkit-dev"`

The CMakeLists.txt:

```
cmake_minimum_required(VERSION 3.16)

project(toolkit_test LANGUAGES CUDA CXX)

find_package(CUDAToolkit REQUIRED)
```

I have the following erro when running cmake:

```
-- Toolchain file defaulted to '/opt/tegra/4.0.6+snapshot/sysroots/x86_64-tdsdk-linux/usr/share/cmake/OEToolchainConfig.cmake'
-- The CUDA compiler identification is NVIDIA 11.4.315
-- The CXX compiler identification is GNU 11.3.0
-- Detecting CUDA compiler ABI info
-- Detecting CUDA compiler ABI info - done
-- Check for working CUDA compiler: /opt/tegra/4.0.6+snapshot/sysroots/x86_64-tdsdk-linux/usr/local/cuda-11.4/bin/nvcc - skipped
-- Detecting CUDA compile features
-- Detecting CUDA compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /opt/tegra/4.0.6+snapshot/sysroots/x86_64-tdsdk-linux/usr/bin/aarch64-oe4t-linux/aarch64-oe4t-linux-g++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Unable to find cuda_runtime.h in "/opt/tegra/4.0.6+snapshot/sysroots/x86_64-tdsdk-linux/usr/local/cuda-11.4/include" for CUDAToolkit_INCLUDE_DIR.
-- Unable to find cublas_v2.h in either "" or "/opt/tegra/4.0.6+snapshot/sysroots/x86_64-tdsdk-linux/usr/math_libs/include"
-- Unable to find cudart library.
CMake Error at /opt/tegra/4.0.6+snapshot/sysroots/x86_64-tdsdk-linux/usr/share/cmake-3.22/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
  Could NOT find CUDAToolkit (missing: CUDAToolkit_INCLUDE_DIR CUDA_CUDART)
  (found version "11.4.315")
Call Stack (most recent call first):
  /opt/tegra/4.0.6+snapshot/sysroots/x86_64-tdsdk-linux/usr/share/cmake-3.22/Modules/FindPackageHandleStandardArgs.cmake:594 (_FPHSA_FAILURE_MESSAGE)
  /opt/tegra/4.0.6+snapshot/sysroots/x86_64-tdsdk-linux/usr/share/cmake-3.22/Modules/FindCUDAToolkit.cmake:814 (find_package_handle_standard_args)
  CMakeLists.txt:15 (find_package)

-- Configuring incomplete, errors occurred!
```

FindCudaToolkit expects the header in `/opt/tegra/4.0.6+snapshot/sysroots/x86_64-tdsdk-linux/usr/local/cuda-11.4/include` but the header is in` /opt/tegra/4.0.6+snapshot/sysroots/armv8a-oe4t-linux//usr/local/cuda-11.4/include`

**Comment by ichergui** [comment]
Hey @hannes09 
You need to adapt your cmake file to make it compatible with the OE build system.

**Comment by madisongh** [comment]
You shouldn't need the `find_package(CUDAToolkit)` if you have `CUDA` listed in the project's languages.  Can you get your project to build if you omit that line?

**Comment by hannes09** [comment]
When I remove `find_package(CUDAToolkit)`  I get the following erro when trying to link against  `cublas_static`:

```
CMake Error at CMakeLists.txt:37 (add_library):
  Target "test" links to target
  "CUDA::cublas_static" but the target was not found.  Perhaps a
  find_package() call is missing for an IMPORTED target, or an ALIAS target
  is missing?
```

**Reply by madisongh** [reply]
One way to do this is with

```
find_package(PkgConfig)
pkg_check_modules(CUBLAS REQUIRED IMPORTED_TARGET)
```
and then use `target_link_libraries(target-name PkgConfig::CUBLAS)` to have your code compile/link with that (and cudart and other libraries).

But in reading through the docs on `FindCUDAToolkit,` it is supposed to locate all that stuff for you, and creates those `CUDA::` imports for the libraries.  I'll see if there's an easy way to make that work for our host/target split sysroot environment.

**Reply by madisongh** [reply]
I tried this out, and it looks like `FindCUDAToolkit` works OK in a bitbake build, but not in an SDK... a couple of additions to the `cuda_toolchain.cmake` file that we install (via the nativesdk-cuda-environment recipe) appear to help:
```
set(CUDAToolkit_TARGET_DIR "${CUDA_TOOLKIT_TARGET_DIR}" CACHE PATH "" FORCE)
find_library(CUDA_CUDART NAMES cudart HINTS ${CUDA_TOOLKIT_TARGET_DIR} PATH_SUFFIXES lib64 lib)
```

**Reply by madisongh** [reply]
I've tested this out and have updated the master, kirkstone, and kirkstone-l4t-r32.7.x branches with the above change, along with a fix for another inconsistency I found between the bitbake environment and the SDK environment.

**Reply by ichergui** [reply]
Thanks @madisongh
